src/wrappers/themis/themispp/secure_cell_seal.hpp

```diff
@@ -0,0 +1,555 @@
+    /// @{
+
+    /**
+     * Encrypt data.
+     *
+     * @param [in] plaintext    non-empty data to be encrypted
+     *
+     * STL-compatible contiguous containers of `uint8_t` and fixed-size C arrays
+     * are supported as input. Plaintext must not be empty.
+     *
+     * Data is encrypted and authentication token is appended to form a single sealed buffer.
+     *
+     * @see encrypt(const Plaintext& plaintext, const Context& context) const
+     *
+     * This method is equivalent to using an empty associated context.
+     *
+     * @returns Newly allocated containter with encrypted data and authentication token.
+     *
+     * @throws themispp::exception_t if `plaintext` is empty.
+     * @throws themispp::exception_t on encryption failure.
+     */
+    template <typename Plaintext>
+    output_data encrypt(const Plaintext& plaintext) const
+    {
+        return encrypt(input_bytes(plaintext), input_buffer());
+    }
+
+    /**
+     * Encrypt data with context.
+     *
+     * @param [in] plaintext    non-empty data to be encrypted
+     * @param [in] context      optional associated context, may be omitted
+     *
+     * STL-compatible contiguous containers of `uint8_t` and fixed-size C arrays
+     * are supported as input. Plaintext must not be empty.
+     *
+     * Data is encrypted and authentication token is appended to form a single sealed buffer.
+     * The context, if provided, is cryptographically mixed with the data, but it not included
+     * into the resulting encrypted message. You will have to provide the same context again
+     * during decryption. Usually this is some plaintext data associated associated with the
+     * encrypted data, such as database row number, protocol message ID, etc.
+     *
+     * @returns Newly allocated containter with encrypted data and authentication token.
+     *
+     * @throws themispp::exception_t if `plaintext` is empty.
+     * @throws themispp::exception_t on encryption failure.
+     */
+    template <typename Plaintext, typename Context>
+    output_data encrypt(const Plaintext& plaintext, const Context& context) const
+    {
+        return encrypt(input_bytes(plaintext), input_bytes(context));
+    }
+
+    /// @}
+    /// @name Data decryption
+    /// @{
+
+    /**
+     * Decrypt data.
+     *
+     * @param [in] encrypted    data to be decrypted
+     *
+     * STL-compatible contiguous containers of `uint8_t` and fixed-size C arrays
+     * are supported as input. Encrypted data cannot be empty.
+     *
+     * Secure Cell validates association with the context data, decrypts the message,
+     * and verifies data integrity using authentication data embedded into the message.
+     *
+     * @see decrypt(const Encrypted& encrypted, const Context& context) const
+     *
+     * This method is equivalent to using an empty associated context.
+     *
+     * @returns Newly allocated containter with decrypted data if everything goes well.
```

Comment:
if we start to use vector types with defined allocator, maybe we should provide some api to provide allocators for results, what do you think?

Reply:
> maybe we should provide some api to provide allocators for results

There is one: that `Allocator` template parameter on Secure Cell templates.

Secure Cell accepts containers with any allocator as input. For output allocations it always uses the specified allocator. Normally this is the standard one (`std::allocator<uint8_t>`) but if something different is needed, users can do

```c++
using my_secure_cell = themispp::impl::secure_cell_seal_with_key<my_allocator<uint8_t>>;

auto cell = my_secure_cell(master_key);

auto encrypted = cell.encrypt(something);
// returns std::vector<uint8_t, my_allocator<uint8_t>>
```

Secure Cell will use `my_allocator` for all non-stack allocations in this case.

Reply:
Oh, I missed that `encrypted` is type of `output_data` which is typedef for vector with generic parameter for allocator.